File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/service/ShareJournalService.java
package com.example.tradelog.api.service;

import com.example.tradelog.api.converter.SyntheticSharesGenerator;
import com.example.tradelog.api.converter.TradeSummaryListConverter;
import com.example.tradelog.api.repository.SharesJournalRepository;
import com.example.tradelog.api.spec.model.ShareJournalModel;
import com.example.tradelog.api.spec.model.TradeSummaryModel;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Optional;


@Service
public class ShareJournalService {

    private SharesJournalRepository sharesJournalRepository;

    public ShareJournalService(SharesJournalRepository sharesJournalRepository) {
        this.sharesJournalRepository = sharesJournalRepository;
    }


    /**
     * Returns all shares per symbol including a calculated synthetic one
     * @param accountId -
     * @param symbol -
     * @return list
     */
    public List<ShareJournalModel> getAllBySymbol(String accountId, String symbol) {
        List<ShareJournalModel> shareJournalModelList = sharesJournalRepository.getAllBySymbol(accountId, symbol);
        shareJournalModelList.addAll(SyntheticSharesGenerator.createSynthetic.apply(shareJournalModelList));
        return shareJournalModelList;
    }


    public Optional<ShareJournalModel> createRecord(String transactionId, ShareJournalModel model) {
        sharesJournalRepository.createShareRecord(transactionId, model);
        return sharesJournalRepository.getByTransactionId(transactionId);
    }

    public boolean deleteRecord(String transactionId) {
        return sharesJournalRepository.deleteRecord(transactionId);
    }

    public Map<String, TradeSummaryModel> getSummaries(String accountId) {
        List<TradeSummaryModel> modelList = sharesJournalRepository.getSummaries(accountId);
        return TradeSummaryListConverter.toMap.apply(modelList);
    }
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/core/service/ShareJournalService.kt
package com.example.tradelog.api.core.service

import arrow.core.Either
import arrow.core.flatMap
import com.example.common.service.ServiceError
import com.example.common.service.toServiceError
import com.example.tradelog.api.core.converter.TradeSummaryUtil
import com.example.tradelog.api.core.model.ShareJournalModel
import com.example.tradelog.api.core.model.TradeSummaryModel
import com.example.tradelog.api.db.repository.ShareJournalRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class ShareJournalService(private val repository: ShareJournalRepository): JournalService<ShareJournalModel> {

    override fun getSummaries(accountId: UUID): Either<ServiceError, Map<String, TradeSummaryModel>> {
        return repository.getSummaries(accountId)
                .mapLeft(::toServiceError)
                .map { TradeSummaryUtil.toMap(it) }
    }

    override fun getAllBySymbol(accountId: UUID, portfolioId: UUID, symbol: String): Either<ServiceError, List<ShareJournalModel>> {
        return repository.getAllBySymbol(accountId, portfolioId, symbol)
                .mapLeft(::toServiceError)
    }

    override fun createRecord(transactionId: UUID, model: ShareJournalModel): Either<ServiceError, ShareJournalModel> {
        val createRecord = repository.createRecord(transactionId, model)
                .mapLeft(::toServiceError)

        val readRecord: (Unit) -> Either<ServiceError, ShareJournalModel> = {
            repository.getById(model.transactionDetails.accountId, transactionId)
                    .mapLeft(::toServiceError)
        }

        return createRecord
                .flatMap(readRecord)
    }

    override fun editRecord(model: ShareJournalModel): Either<ServiceError, Unit> {
        return repository.editRecord(model)
                .mapLeft(::toServiceError)
    }

    override fun deleteRecord(transactionId: UUID): Either<ServiceError, Unit> {
        return repository.deleteRecord(transactionId)
                .mapLeft(::toServiceError)
    }

    override fun getById(accountId: UUID, transactionId: UUID): Either<ServiceError, ShareJournalModel> {
        return repository.getById(accountId, transactionId)
                .mapLeft(::toServiceError)
    }
}

